windows/RNPrint/RNPrint.h

```diff
@@ -0,0 +1,402 @@
+#pragma once
+
+#include "pch.h"
+#include "NativeModules.h"
+
+#include <functional>
+#include <sstream>
+#include <mutex>
+
+#include "JSValue.h"
+
+using namespace winrt::Microsoft::ReactNative;
+using namespace winrt::Windows::Data::Pdf;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+using namespace winrt::Windows::Web::Http;
+using namespace winrt::Windows::Graphics::Printing;
+using namespace winrt::Windows::Graphics::Printing::OptionDetails;
+using namespace winrt::Windows::UI::Xaml::Printing;
+using namespace winrt::Windows::UI::Xaml::Controls;
+using namespace winrt::Windows::UI::Xaml::Media::Imaging;
+using namespace winrt::Windows::UI::Xaml::Media;
+
+
+namespace winrt::RNPrint
+{
+  // Represent the options object passed to print()
+  struct RNPrintOptions
+  {
+    std::optional<std::string> html = std::nullopt;
+    std::optional<std::string> filePath = std::nullopt;
+    bool isLandscape = false;
+    std::string jobName;
+  };
+
+  REACT_MODULE(RNPrint);
+  struct RNPrint
+  {
+    const std::string Name = "RNPrint";
+
+    // Default name for the print job.
+    inline static const std::string defaultJobName = "Document";
+
+    // Name for the XAML Canvas to add preview folders to.
+    inline static const std::string RNPrintCanvasName = "RNPrintCanvas";
+
+    // These are needed between print PDF callbacks, so we place them in the struct.
+    ReactContext reactContext = nullptr;
+    Canvas printCanvas = nullptr; // holds reference to RNPrintCanvas in visual tree
+    PrintPageDescription printPageDescr;
+    PrintDocument printDoc = nullptr;
+    IPrintDocumentSource printDocSource;
+    int numberOfPages = 0;
+    std::map<int, std::optional<xaml::UIElement>> pageCollection; // references to the xaml pages
```

Comment:
it doesn't seem like you need this to be a map, you can just use std::vector

Reply:
These can be generated out of order and a map also would support a future feature to allow page range printing.
I think a map works better here.
Using a map also follows the Windows Universal Samples code for the Printing API more closely.

Reply:
agree that they can be generated out of order, but you are capturing the index i in the lambda, so maybe you can use that? not a big concern here, just looking to simplify since std::map is big and clunky

Reply:
There's also the part of the `when_all` later on which wants to send it to the printer in the right order.
Since we are generating a XAML element for each member in the collection, the `std::map` overhead shouldn't have much impact in comparison.
The collection also isn't expected to end up having many members, since it's one per page.